### PR TITLE
AFA-116 Review task that deletes expired tokens

### PR DESCRIPTION
**Pull request by ecilveks**

### CODE REVIEW

File: oauth2_provider/models.py [hunk 1]
```diff
@@ -434,8 +436,18 @@ def get_refresh_token_model():
     return apps.get_model(oauth2_settings.REFRESH_TOKEN_MODEL)
 
 
-def clear_expired():
-    now = timezone.now()
+def clear_expired(before = None):
```
Comment (komar) [inline oauth2_provider/models.py]:
```suggestion
def clear_expired(before=None):
```

File: oauth2_provider/models.py [hunk 2]
```diff
@@ -434,8 +436,18 @@ def get_refresh_token_model():
...
+    """Clear expired tokens, add custom date to limit clear tokens before this date."""
+    if not before:
+        now = timezone.now()
+    else:
+        now = before
+        if not type(before).__name__ == 'datetime':
```
Comment (komar) [inline oauth2_provider/models.py]: Is this branch needed since management command uses `dateutil.parser.parse`?

**Review by komar — CHANGES_REQUESTED** [review]
Please add tests.

File: oauth2_provider/models.py (hunk 2 above)
Comment (ecilveks) [reply oauth2_provider/models.py]: we need this, because celery task call this function direct, without management command

Comment (komar) [reply oauth2_provider/models.py]: Could Celery task call it with `datetime` rather than a `string`?

Comment (ecilveks) [reply oauth2_provider/models.py]: celery call without extra params, but better is protect this function from incorrect input data

Comment (komar) [reply oauth2_provider/models.py]: Ok, you could check for the type `isinstance`, log if this is not a valid `datetime`and return

File: oauth2_provider/models.py [hunk 3]
```diff
@@ -434,8 +435,15 @@ def get_refresh_token_model():
     return apps.get_model(oauth2_settings.REFRESH_TOKEN_MODEL)
 
 
-def clear_expired():
-    now = timezone.now()
+def clear_expired(before=None):
+    """Clear expired tokens, add custom date to limit clear tokens before this date."""
+    if not before:
```
Comment (komar) [inline oauth2_provider/models.py]: I feel this whole block could be simplified to:
```suggestion
    if not tokens_expired_before_date:
        tokens_expired_before_date = timezone.now()
    if isinstance(tokens_expired_before_date, date) is False:
            logger.info('Not valid datetime provided.')
            return  
```

File: tests/test_models.py [hunk 4]
```diff
@@ -347,3 +349,195 @@ def test_clear_expired_tokens_with_tokens(self):
...
+        self.client.login(username="test_user", password="123456")
+        app = Application.objects.create(
+            name="test_app",
+            redirect_uris="http://localhost http://example.com http://example.org",
+            user=self.user,
+            client_type=Application.CLIENT_CONFIDENTIAL,
+            authorization_grant_type=Application.GRANT_AUTHORIZATION_CODE,
+        )
+        oauth2_settings.REFRESH_TOKEN_EXPIRE_SECONDS = 0
+        AccessToken.objects.create(
```
Comment (komar) [inline tests/test_models.py]:
```suggestion
       access_token1 = AccessToken.objects.create(
```

File: tests/test_models.py [hunk 5]
```diff
@@ -347,3 +349,195 @@ def test_clear_expired_tokens_with_tokens(self):
...
+            token="999",
+            expires=timezone.now()-timedelta(days=30),
+            scope=2,
+            application=app,
+            user=self.user,
+            created=timezone.now()-timedelta(days=40),
+            updated=timezone.now()-timedelta(days=40),
+            )
+        access_token1 = AccessToken.objects.last()
```
Comment (komar) [inline tests/test_models.py]:
```suggestion
```

File: tests/test_models.py [hunk 6]
```diff
@@ -347,3 +349,195 @@ def test_clear_expired_tokens_with_tokens(self):
...
+        RefreshToken.objects.create(
+            token = 'refresh999',
+            access_token = access_token1,
+            application = app,
+            user=self.user,
+            )
+        AccessToken.objects.create(
```
Comment (komar) [inline tests/test_models.py]:
```suggestion
        access_token2  = AccessToken.objects.create(
```

File: tests/test_models.py [hunk 7]
```diff
@@ -347,3 +349,195 @@ def test_clear_expired_tokens_with_tokens(self):
...
+            token="101010",
+            expires=timezone.now()-timedelta(days=15),
+            scope=2,
+            application=app,
+            user=self.user,
+            created=timezone.now()-timedelta(days=20),
+            updated=timezone.now()-timedelta(days=20),
+            )
+        access_token2 = AccessToken.objects.last()
```
Comment (komar) [inline tests/test_models.py]:
```suggestion
```

File: tests/test_models.py [hunk 8]
```diff
@@ -347,3 +349,195 @@ def test_clear_expired_tokens_with_tokens(self):
...
+        self.client.login(username="test_user", password="123456")
+        app = Application.objects.create(
+            name="test_app",
+            redirect_uris="http://localhost http://example.com http://example.org",
+            user=self.user,
+            client_type=Application.CLIENT_CONFIDENTIAL,
+            authorization_grant_type=Application.GRANT_AUTHORIZATION_CODE,
+        )
+        oauth2_settings.REFRESH_TOKEN_EXPIRE_SECONDS = 0
+        AccessToken.objects.create(
```
Comment (komar) [inline tests/test_models.py]:
```suggestion
        access_token1 =  AccessToken.objects.create(
```

File: tests/test_models.py [hunk 9]
```diff
@@ -347,3 +349,195 @@ def test_clear_expired_tokens_with_tokens(self):
...
+            token="999",
+            expires=timezone.now()-timedelta(days=30),
+            scope=2,
+            application=app,
+            user=self.user,
+            created=timezone.now()-timedelta(days=40),
+            updated=timezone.now()-timedelta(days=40),
+            )
+        access_token1 = AccessToken.objects.last()
```
Comment (komar) [inline tests/test_models.py]:
```suggestion
```

File: tests/test_models.py [hunk 10]
```diff
@@ -347,3 +349,195 @@ def test_clear_expired_tokens_with_tokens(self):
...
+        RefreshToken.objects.create(
```
Comment (komar) [inline tests/test_models.py]:
```suggestion
        refresh_token1 = RefreshToken.objects.create(
```

File: tests/test_models.py [hunk 11]
```diff
@@ -347,3 +349,195 @@ def test_clear_expired_tokens_with_tokens(self):
...
+            token = 'refresh999',
+            access_token = access_token1,
+            application = app,
+            user=self.user,
+            created=timezone.now()-timedelta(days=4),
+            updated=timezone.now()-timedelta(days=40),
+            )
+        refresh_token1= RefreshToken.objects.last()
```
Comment (komar) [inline tests/test_models.py]:
```suggestion
```

File: tests/test_models.py [hunk 12]
```diff
@@ -347,3 +349,195 @@ def test_clear_expired_tokens_with_tokens(self):
...
+        AccessToken.objects.create(
```
Comment (komar) [inline tests/test_models.py]:
```suggestion
         access_token2 = AccessToken.objects.create(
```

File: tests/test_models.py [hunk 13]
```diff
@@ -347,3 +349,195 @@ def test_clear_expired_tokens_with_tokens(self):
...
+            token="101010",
+            expires=timezone.now()-timedelta(days=15),
+            scope=2,
+            application=app,
+            user=self.user,
+            created=timezone.now()-timedelta(days=20),
+            updated=timezone.now()-timedelta(days=20),
+            )
+        access_token2 = AccessToken.objects.last()
```
Comment (komar) [inline tests/test_models.py]:
```suggestion
```

File: tests/test_models.py [hunk 14]
```diff
@@ -347,3 +349,195 @@ def test_clear_expired_tokens_with_tokens(self):
...
+        AccessToken.objects.create(
+            token="888",
+            expires=timezone.now()-timedelta(days=15),
+            scope=2,
+            application=app,
+            user=self.user,
+            created=timezone.now()-timedelta(days=20),
+            updated=timezone.now()-timedelta(days=20),
+            )
+        ttokens = AccessToken.objects.count()
```
Comment (komar) [inline tests/test_models.py]: This name is not descriptive
```suggestion
        access_token_count = AccessToken.objects.count()
```
would be better.

The same applies to `expiredt` and `rtokens` in other tests.

I know this is used in other tests, but it not good for reading the code.

**Review by komar — CHANGES_REQUESTED** [review]

**Review by komar — APPROVED** [review]

**Pull request merged**